frontend/cypress/integration/common/sidecar_injection.ts

```diff
@@ -0,0 +1,274 @@
+import { Given, Then, When } from "cypress-cucumber-preprocessor/steps";
+
+Given('a namespace without override configuration for automatic sidecar injection', function () {
+    this.targetNamespace = "default";
+
+    // Make sure that the target namespace does not have override configuration
+    cy.request('PATCH', '/api/namespaces/' + this.targetNamespace, {
+        metadata: {
+            labels: {
+                "istio-injection": null,
+                "istio.io/rev": null
+            }
+        }
+    });
+});
+
+Given('a namespace which has override configuration for automatic sidecar injection', function () {
+    this.targetNamespace = "default";
+    this.istioInjection = "enabled";
+
+    // Make sure that the target namespace has some override configuration
+    cy.request('PATCH', '/api/namespaces/' + this.targetNamespace, {
+        metadata: {
+            labels: {
+                "istio-injection": this.istioInjection,
+                "istio.io/rev": null
+            }
+        }
+    });
+});
+
+Given('the override configuration for sidecar injection is {string}', function (enabledOrDisabled) {
+    if (this.istioInjection !== enabledOrDisabled) {
+        cy.request('PATCH', '/api/namespaces/' + this.targetNamespace, {
+            metadata: {
+                labels: {
+                    "istio-injection": enabledOrDisabled,
+                    "istio.io/rev": null
+                }
+            }
+        });
+        this.istioInjection = enabledOrDisabled;
+    }
+});
+
+Given('a workload without a sidecar', function () {
+    this.targetNamespace = 'default';
+    this.targetWorkload = 'sleep';
+
+    // To achieve a workload without sidecar, we turn off injection in its namespace
+    // and also remove any override annotation for sidecar injection.
+    this.namespaceAutoInjectionEnabled = false;
+    this.workloadHasSidecar = false;
+    this.workloadHasAutoInjectionOverride = false;
+
+    // Make sure that injection in the namespace is turned off
+    cy.request('PATCH', '/api/namespaces/' + this.targetNamespace, {
+        metadata: {
+            labels: {
+                "istio-injection": null,
+                "istio.io/rev": null
+            }
+        }
+    });
+
+    // Make sure that the workload does not have override configuration
+    cy.request('PATCH', `/api/namespaces/${this.targetNamespace}/workloads/${this.targetWorkload}?type=Deployment`, {
+        spec: {
+            template: {
+                metadata: {
+                    labels: {
+                        "sidecar.istio.io/inject": null
+                    }
+                }
+            }
+        }
+    });
+
+    // Restart the workload to ensure changes are applied.
+    cy.exec('kubectl delete pod -n default -l app=sleep')
```

Comment:
Sorry, I'm -1 on this, I don't think cypress should invoke the system commands.
You can prepare the scenario in the cluster in the external scripts (as we did with the alpha/beta namespaces) and ensure that the test operates with the UI with the state of the cluster.

Reply:
Maybe not here but I think we'll eventually need the tests to execute some system commands. A good example would be the wizards. Without doing a `kubectl get <objecttype>`, there's not an easy way to fetch an object that a wizard created. We could just test the kiali api response object but that doesn't test that the actual kube object was created/edited properly. `kubectl` and/or `oc` are already prerequisites to running the UI tests.

Reply:
I don't think we should validate things out of the UI unless there is a good reason for that.
The wizards validate the Istio config created from the UI.

There could be always exceptions, but there is a suite of e2e that is designed for this scope.

I have the feeling that we are mixing it here and then it would add more complexity in the future.

Reply:
This is a `Given` step.
In the Gherkin reference, it is stated that the purpose of a `Given` is simply "_to put the system the system in a known state_". It is even mentioned that we could add data to a test database (but we don't have a DB, so that does not apply).

IMHO, the docs give the idea that `Given` steps aren't doing testing. So, I understand that it should be OK to use whatever is _at hand_ to achieve the pre-conditions. I know I'm using the API and that is not with the purpose of testing the API, but to achieve the system state as described in the sentence. Indeed, here I'm assuming the API just works (no assertions are done).

I'm also invoking some CLI tools, with that same purpose. Here, I'm using `kubectl` because, sometimes the pod does not restart automatically and, thus, desired state is not reached properly without this restart.

I know these `Given` steps may be different to what we've been writing for the other tests. But, notice that these tests are probably the first ones that are doing changes to the environment as part of testing (i.e. injecting/removing sidecars via changing annotations). If I were to purely use the UI to also attain the pre-conditions, I would end up in a chicken-egg situation, because for some `Scenario`s I would need to use UI features and assume that they "just work" and then test those same UI features in another `Scenario`s (so, if something actually breaks in the UI, tests would break in pairs: one where preconditions fail to be achieved, and the other when actual assertions would fail).

So, I thought this was the best way to write the tests: using whatever is _at hand_ to achieve pre-conditions to then run the desired and relevant UI verifications. Indeed, I invite you to look at the `Then` and `When` steps. Those ones are implemented only with UI manipulations.

Reply:
If "Given" statements are used for preparing the state of the test, it sounds reasonable.
But yes, the test should be UI only, unless there are good reasons to find a workaround (i.e. graph area).

I wanted to avoid situations where a test creates some configuration and the test uses the private API to check something instead to navigate to the proper page (because it's easier to invoke that private API than navigate in the UI and build a test).

As soon as the commands are valid in any environment (upstream/downstream), it sounds reasonable.

I'd suggest adding comments in the "Given" area making clear that those API invocations are private, otherwise, we'll have these discussions on every new test, or we may have a variety of styles for test writings and it's the moment to try align them.